File: 백준/Gold/17142. 연구소 3/연구소 3.py
import sys

input = sys.stdin.readline

from itertools import combinations
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

n, m = map(int, input().split())

arr = [list(map(int, input().split())) for _ in range(n)]

visited = [[0] * n for _ in range(n)]

q = deque()
virus = []
wall = []
for i in range(n):
    for j in range(n):
        if arr[i][j] == 2:
            virus.append((i, j))
        if arr[i][j] == 1:
            wall.append((i, j))

allcase = list(combinations(virus, m))
inf = int(1e9)
result = inf
for case in allcase:
    flag = True
    q = deque(case)
    visited = [[0] * n for _ in range(n)]
    for c in case:
        visited[c[0]][c[1]] = 1

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < n:
                if arr[nx][ny] == 0 and visited[nx][ny] == 0:
                    q.append((nx, ny))
                    visited[nx][ny] = visited[x][y] + 1
                # 비활성화된 바이러스
                if arr[nx][ny] == 2 and visited[nx][ny] == 0:
                    q.append((nx, ny))
                    visited[nx][ny] = visited[x][y] + 1

    max_count = 0
    for i in range(n):
        for j in range(n):
            if visited[i][j] == 0 and arr[i][j] == 0:
                flag = False
            if visited[i][j] > 0 and visited[i][j] > max_count and arr[i][j] == 0:
                max_count = visited[i][j]
        if flag == False:
            break
    if flag:
        result = min(result,max_count)
    # print(result)?

if result == inf:
    print(-1)
else:
    if result == 0:
        print(0)
    else:
        print(result-1)

File: 백준/Gold/17142. 연구소 3/연구소 3.js
const fs = require("fs");

const input =
    process.platform === "linux"
        ? fs.readFileSync("/dev/stdin").toString().split("\n")
        : fs.readFileSync("example.txt").toString().split("\r\n");

const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

const [num, ...a] = input;
const [n, m] = num.split(" ").map(Number);
const arr = a.map((e) => e.split(" ").map(Number));

const virus = [];
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
        if (arr[i][j] === 2) {
            virus.push([i, j]);
        }
    }
}

const inf = 1e9;
let result = inf;

const bfs = (selectedViruses) => {
    const visited = Array.from(Array(n), () => Array(n).fill(0));
    const q = [];
    
    selectedViruses.forEach(([x, y]) => {
        q.push([x, y, 0]); // (x, y, time)
        visited[x][y] = 1;
    });

    let maxTime = 0;

    while (q.length > 0) {
        const [x, y, time] = q.shift();

        for (let i = 0; i < 4; i++) {
            const nx = x + dx[i];
            const ny = y + dy[i];

            if (nx >= 0 && nx < n && ny >= 0 && ny < n && visited[nx][ny] === 0 && arr[nx][ny] !== 1) {
                visited[nx][ny] = 1;
                q.push([nx, ny, time + 1]);
                if (arr[nx][ny] === 0) {
                    maxTime = time + 1;
                }
            }
        }
    }

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (arr[i][j] === 0 && visited[i][j] === 0) {
                return;
            }
        }
    }
    result = Math.min(result, maxTime);
}

const dfs = (start, count, selectedViruses) => {
    if (count === m) {
        bfs(selectedViruses);
        return;
    }

    for (let i = start; i < virus.length; i++) {
        selectedViruses.push(virus[i]);
        dfs(i + 1, count + 1, selectedViruses);
        selectedViruses.pop();
    }
}

dfs(0, 0, []);

console.log(result === inf ? -1 : result);